algorithms task (5)/student/include/impl.h

```diff
@@ -10,4 +13,19 @@
  * @return gathered range
  */
 template <typename BiIt, typename UnaryPredicate>
-std::pair<BiIt, BiIt> gather(BiIt begin, BiIt end, BiIt position, UnaryPredicate predicate);
+std::pair<BiIt, BiIt> gather(BiIt begin, BiIt end, BiIt position, UnaryPredicate predicate)
+{
+    const auto notPredicate = [predicate](const typename std::iterator_traits<BiIt>::reference value)
+    {
+        return !predicate(value);
+    };
+
+    auto partitionBegin = std::partition(begin, std::next(position), notPredicate);
+    auto partitionEnd = std::partition(position, end, predicate);
+
+    if (partitionBegin == std::next(position))
```

Comment:
What if position is end?

Reply:
Undefined behavior, but I assumed that it doesn't make sense to pass end() as position. I think it is not so important here. Do you think it should be fixed?

Reply:
consider the empty range

Reply:
Fixed.